File: backend/app/services/file_roles.py
from __future__ import annotations

from fnmatch import fnmatch
from pathlib import PurePosixPath

from backend.app.services.graph.models import CodeGraphEdge, CodeGraphNode

LOCKFILE_NAMES = {"uv.lock", "package-lock.json", "pnpm-lock.yaml", "yarn.lock"}
GENERATED_DIR_NAMES = {
    ".next",
    ".nuxt",
    ".svelte-kit",
    "build",
    "coverage",
    "dist",
    "htmlcov",
    "out",
    "target",
}
GENERATED_FILE_SUFFIXES = {
    ".bundle.js",
    ".bundle.css",
    ".d.ts",
    ".generated.py",
    ".generated.ts",
    ".generated.tsx",
    ".min.css",
    ".min.js",
}
GENERATED_FILE_NAMES = LOCKFILE_NAMES
VENDOR_DIR_NAMES = {
    ".git",
    ".hg",
    ".svn",
    ".venv",
    "node_modules",
    "site-packages",
    "vendor",
    "vendors",
    "venv",
}
TEST_DIR_NAMES = {"__tests__", "e2e", "spec", "specs", "test", "tests"}


def normalize_file_path(file_path: str) -> str:
    return file_path.replace("\\", "/").strip("/")


def is_test_file(file_path: str, test_glob: str | None = None) -> bool:
    normalized = normalize_file_path(file_path)
    lowered = normalized.lower()
    if test_glob and fnmatch(normalized, test_glob):
        return True
    name = lowered.rsplit("/", 1)[-1]
    parts = set(PurePosixPath(lowered).parts)
    return (
        name.startswith("test_")
        or name.endswith("_test.py")
        or name.endswith("_test.go")
        or ".test." in name
        or ".spec." in name
        or bool(parts & TEST_DIR_NAMES)
    )


def is_generated_file(file_path: str) -> bool:
    normalized = normalize_file_path(file_path)
    lowered = normalized.lower()
    name = lowered.rsplit("/", 1)[-1]
    parts = set(PurePosixPath(lowered).parts)
    return (
        name in GENERATED_FILE_NAMES
        or any(name.endswith(suffix) for suffix in GENERATED_FILE_SUFFIXES)
        or bool(parts & GENERATED_DIR_NAMES)
    )


def is_vendor_file(file_path: str) -> bool:
    parts = set(PurePosixPath(normalize_file_path(file_path).lower()).parts)
    return bool(parts & VENDOR_DIR_NAMES)


def is_wiki_noise_file(file_path: str) -> bool:
    return is_test_file(file_path) or is_generated_file(file_path) or is_vendor_file(file_path)


def is_wiki_noise_node(node: CodeGraphNode) -> bool:
    if node.metadata.get("external"):
        return True
    return bool(node.file_path and is_wiki_noise_file(node.file_path))


def filter_wiki_graph(
    nodes: list[CodeGraphNode],
    edges: list[CodeGraphEdge],
) -> tuple[list[CodeGraphNode], list[CodeGraphEdge]]:
    filtered_nodes = [node for node in nodes if not is_wiki_noise_node(node)]
    node_ids = {node.id for node in filtered_nodes}
    filtered_edges = [
        edge for edge in edges if edge.source_id in node_ids and edge.target_id in node_ids
    ]
    return filtered_nodes, filtered_edges

File: backend/src/services/fileRoles.ts
import type { CodeGraphEdge, CodeGraphNode } from "../types.js";

const LOCKFILE_NAMES = new Set([
  "uv.lock",
  "package-lock.json",
  "pnpm-lock.yaml",
  "yarn.lock",
]);

const GENERATED_DIR_NAMES = new Set([
  ".next",
  ".nuxt",
  ".svelte-kit",
  "build",
  "coverage",
  "dist",
  "htmlcov",
  "out",
  "target",
]);

const GENERATED_FILE_SUFFIXES = [
  ".bundle.js",
  ".bundle.css",
  ".d.ts",
  ".generated.py",
  ".generated.ts",
  ".generated.tsx",
  ".min.css",
  ".min.js",
];

const VENDOR_DIR_NAMES = new Set([
  ".git",
  ".hg",
  ".svn",
  ".venv",
  "node_modules",
  "site-packages",
  "vendor",
  "vendors",
  "venv",
]);

const TEST_DIR_NAMES = new Set([
  "__tests__",
  "e2e",
  "spec",
  "specs",
  "test",
  "tests",
]);

export function normalizeFilePath(filePath: string): string {
  return filePath
    .replace(/\\/g, "/")
    .replace(/^\/+|\/+$/g, "")
    .trim();
}

export function isTestFile(filePath: string): boolean {
  const normalized = normalizeFilePath(filePath).toLowerCase();
  const name = normalized.split("/").pop() ?? "";
  const parts = new Set(normalized.split("/").filter(Boolean));
  return (
    name.startsWith("test_") ||
    name.endsWith("_test.py") ||
    name.endsWith("_test.go") ||
    name.includes(".test.") ||
    name.includes(".spec.") ||
    intersects(parts, TEST_DIR_NAMES)
  );
}

export function isGeneratedFile(filePath: string): boolean {
  const normalized = normalizeFilePath(filePath).toLowerCase();
  const name = normalized.split("/").pop() ?? "";
  const parts = new Set(normalized.split("/").filter(Boolean));
  return (
    LOCKFILE_NAMES.has(name) ||
    GENERATED_FILE_SUFFIXES.some((suffix) => name.endsWith(suffix)) ||
    intersects(parts, GENERATED_DIR_NAMES)
  );
}

export function isVendorFile(filePath: string): boolean {
  return intersects(
    new Set(
      normalizeFilePath(filePath).toLowerCase().split("/").filter(Boolean),
    ),
    VENDOR_DIR_NAMES,
  );
}

export function isWikiNoiseFile(filePath: string): boolean {
  return (
    isTestFile(filePath) || isGeneratedFile(filePath) || isVendorFile(filePath)
  );
}

export function isWikiNoiseNode(node: CodeGraphNode): boolean {
  if (node.metadata.external === true) {
    return true;
  }
  return Boolean(node.file_path && isWikiNoiseFile(node.file_path));
}

export function filterWikiGraph(
  nodes: CodeGraphNode[],
  edges: CodeGraphEdge[],
): { nodes: CodeGraphNode[]; edges: CodeGraphEdge[] } {
  const filteredNodes = nodes.filter((node) => !isWikiNoiseNode(node));
  const nodeIds = new Set(filteredNodes.map((node) => node.id));
  return {
    nodes: filteredNodes,
    edges: edges.filter(
      (edge) => nodeIds.has(edge.source_id) && nodeIds.has(edge.target_id),
    ),
  };
}

function intersects(left: Set<string>, right: Set<string>): boolean {
  for (const value of left) {
    if (right.has(value)) {
      return true;
    }
  }
  return false;
}
